winforms/src/toga_winforms/widgets/scrollcontainer.py

```diff
@@ -1,87 +1,123 @@
+from travertino.node import Node
 
-from .base import Widget
+from toga_winforms.container import Container
 
+from .base import Widget
 
-class ScrollContainer(Widget):
+# On Windows, scroll bars usually appear only when the content is larger than the
+# container. However, this complicates layout. For example, if the content fits in
+# the container horizontally but is taller vertically, we then need to do a second
+# layout pass with a slightly narrower horizontal size to account for the vertical
+# scroll bar (https://stackoverflow.com/questions/28418026).
+#
+# A previous attempt to avoid this, by making the scroll bars always visible, was not
+# successful (see commit "Initial Winforms implementation" on 2023-07-10). Although the
+# ScrollableControl API does provide Visible and Enabled properties for each scroll bar,
+# they behave in confusing and undocumented ways, e.g.:
+#
+#  * https://stackoverflow.com/questions/8690643
+#  * https://stackoverflow.com/questions/5489273
+#
+# So the current implementation just uses the default AutoScroll behavior, and does a
+# second layout pass where necessary.
+
+
+class ScrollContainer(Widget, Container):
     def create(self):
-        self.native = WinForms.Panel()
-        self.native.interface = self.interface
+        self.native = Panel()
         self.native.AutoScroll = True
+        self.init_container(self.native)
+
+        # The Scroll event only fires on direct interaction with the scroll bar. It
+        # doesn't fire when using the mouse wheel, and it doesn't fire when setting
+        # AutoScrollPosition either, despite the documentation saying otherwise.
         self.native.Scroll += self.winforms_scroll
         self.native.MouseWheel += self.winforms_scroll
 
     def winforms_scroll(self, sender, event):
-        if self.interface.on_scroll is not None:
-            self.interface.on_scroll(self.interface)
-
-    def set_content(self, widget):
-        self.inner_container = widget
-
-        widget.viewport = WinFormsViewport(self.native, self)
-        widget.frame = self
-
-        for child in widget.interface.children:
-            child._impl.container = widget
-
-        self.native.Controls.Add(self.inner_container.native)
+        self.interface.on_scroll(None)
+
+    def set_bounds(self, x, y, width, height):
+        super().set_bounds(x, y, width, height)
+        self.resize_content(width, height)
+
+    def refreshed(self):
+        full_width, full_height = (self.width, self.height)
+        inset_width = full_width - SystemInformation.VerticalScrollBarWidth
+        inset_height = full_height - SystemInformation.HorizontalScrollBarHeight
+        layout = self.interface.content.layout
+
+        # Temporarily reduce container size to account for scroll bars (see explanation
+        # at the top of this file).
+        def apply_insets():
+            need_scrollbar = False
+            if self.vertical and layout.height > self.height:
```

Comment:
Should this comparison be with self.height, or inset_height?

Reply:
`self.height` is the correct value in both cases:

* In the first call to `apply_insets`, `self.height == full_height`, because you're checking whether the initial layout needs a vertical scroll bar.
* In the second call, `self.height` may have been set to `inset_height` if a horizontal scroll bar was added by the first call, in which case we now need to check whether the second layout fits in that smaller space.